app/src/editor/canvas/index.jsx

```diff
@@ -172,15 +172,52 @@ const CanvasEditComponent = class CanvasEdit extends Component {
     }
 
     loadNewDefinition = async (hash) => {
-        const { replace } = this.props
+        const { replace, undo } = this.props
         const module = CanvasState.getModule(this.props.canvas, hash)
-        const newModule = await sharedServices.getModule(module)
+        try {
+            const newModule = await sharedServices.getModule(module)
 
-        if (!this.unmounted) {
-            replace(() => CanvasState.updateModule(this.props.canvas, hash, () => newModule))
+            if (!this.unmounted) {
+                replace(() => CanvasState.updateModule(this.props.canvas, hash, () => newModule))
+            }
+        } catch (error) {
+            console.error(error.message)
+            // undo value change
+            undo()
         }
     }
 
+    reloadModule = async (hash) => {
+        const { canvas, apiKey } = this.props
+        const module = CanvasState.getModule(canvas, hash)
+
+        const oldParamNames = module.params.map((param) => param.name)
+        const oldInputNames = module.inputs.map((input) => input.name)
+        const oldOutputNames = module.outputs.map((output) => output.name)
+
+        const res = await sharedServices.send({
+            canvasId: canvas.id,
+            moduleHash: module.hash,
+            data: {
+                type: 'json',
+            },
+            apiKey,
+        })
+
+        if (this.unmounted) { return }
+
+        const { json: newModule } = res
+
+        this.setCanvas({ type: 'Reload Module' }, (canvas) => (
```

Comment:
This creates an undo stack entry. Is that correct?

Reply:
This was actually old code, so ignore.